lib/storage/gcs.js

```diff
@@ -0,0 +1,152 @@
+          var defaultSecure = ((!options.port) || (options.port === 443));
+          var secure = options.secure || defaultSecure;
+          var port = options.port || 443;
+          var protocol = secure ? 'https://' : 'http://';
+          if (secure && (port === 443)) {
+            port = '';
+          } else if ((!secure) && (port === 80)) {
+            port = '';
+          } else {
+            port = ':' + port;
+          }
+          endpoint = protocol + endpoint + port;
+        }
+      }
+      // The storage client auth relies on the presence of the service account
+      // file path expressed in the environment variable
+      // GOOGLE_APPLICATION_CREDENTIALS and, of course, the presence of such file.
+      //
+      //
+      // See https://cloud.google.com/docs/authentication/getting-started
+      client = new Storage();
+      bucketName = options.bucket;
+      defaultTypes = require(__dirname + "/contentTypes.js");
+      if (options.contentTypes) {
+        _.extend(contentTypes, defaultTypes, options.contentTypes);
+      } else {
+        contentTypes = defaultTypes;
+      }
+      defaultTypes = require(__dirname + "/contentTypes.js");
+      if (options.contentTypes) {
+        _.extend(contentTypes, defaultTypes, options.contentTypes);
+      } else {
+        contentTypes = defaultTypes;
+      }
+      https = options.https;
+      cachingTime = options.cachingTime;
+      return callback(null);
+    },
+
+    copyIn: function(localPath, path, options, callback) {
+      async.waterfall([
+        function(callbackOne) {
+          var cleanPath = cleanKey(path);
+          var ext = extname(cleanPath);
+          if (ext.length) {
+            ext = ext.substr(1);
+          }
+          var contentType = contentTypes[ext];
+          if (!contentType) {
+            console.log(`unknown extension ${ext}, using mime octet-stream`);
+            contentType = 'application/octet-stream';
+          }
+
+          let cacheControl = 'no-cache';
+          if (cachingTime) {
+            cacheControl = 'public, max-age=' + cachingTime;
+          }
+          callbackOne(null, {
+            destination: cleanPath,
+            gzip: true,
+            metadata: {
+              cacheControl: cacheControl,
+              ContentType: contentType
+            }
+          });
+        },
+        function(uploadOptions, callbackTwo) {
+          client.bucket(bucketName).upload(localPath, uploadOptions).then(
+            function(apiResponse) {
+              // set the asset to public by default.
+              var partOne = apiResponse[0];
+              client.bucket(bucketName).file(partOne.metadata.name).makePublic().then(function (__) {
+                callbackTwo(null, apiResponse);
+              }, function (err) {
+                callback(err);
+              });
+            }).catch(function (reason) {
+            callback(reason);
+          });
+        },
+        function(apiResponse, __) {
```

Comment:
This works, but it should have its own callback, and the final `callback` should be passed as the final argument to `async.waterfall`, for consistency with the expectations of folks familiar with the module.

Reply:
I don't mind cleaning this up.

Reply:
Oops, you're right. I can fix this as I do the other changes you requested :)